### PR TITLE
[SPARK-25699][SQL] Partially push down conjunctive predicated in ORC

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

Inspired by https://github.com/apache/spark/pull/22574 .
We can partially push down top level conjunctive predicates to Orc.
This PR improves Orc predicate push down in both SQL and Hive module.

## How was this patch tested?

New unit test.

### CODE REVIEW

**Comment by gengliangwang** [comment]
@dbtsai @gatorsmile

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala [hunk 1]
```diff
@@ -382,5 +382,40 @@ class OrcFilterSuite extends OrcTest with SharedSQLContext {
...
+    assertResult("leaf-0 = (LESS_THAN a 10), expr = leaf-0") {
+      OrcFilters.createFilter(schema, Array(
+        And(
+          LessThan("a", 10),
+          StringContains("b", "prefix")
+        )
+      )).get.toString
+    }
+
+    // Safely remove unsupported `StringContains` predicate and push down `LessThan`
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala]: There is another predicate pushed down. We shall mention it too?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala [hunk 2]
```diff
@@ -138,39 +138,75 @@ private[sql] object OrcFilters {
...
-          rhs <- buildSearchArgument(dataTypeMap, right, lhs)
-        } yield rhs.end()
+        //
+        // Pushing one side of AND down is only safe to do at the top level or in the child
+        // AND before hitting NOT or OR conditions, and in this case, the unsupported predicate
+        // can be safely removed.
+        val leftBuilderOption = createBuilder(dataTypeMap, left,
+          newBuilder, canPartialPushDownConjuncts)
+        val rightBuilderOption =
+          createBuilder(dataTypeMap, right, newBuilder, canPartialPushDownConjuncts)
```
Comment (dbtsai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala]: Can you make the format the same as `leftBuilderOption`? Also, add another empty line before `(leftBuilderOption, rightBuilderOption)`. Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala [hunk 3]
```diff
@@ -138,39 +138,75 @@ private[sql] object OrcFilters {
...
+            builder.startOr(), canPartialPushDownConjuncts = false)
+          rhs <- createBuilder(dataTypeMap, right, lhs, canPartialPushDownConjuncts = false)
         } yield rhs.end()
 
       case Not(child) =>
         for {
-          _ <- buildSearchArgument(dataTypeMap, child, newBuilder)
-          negate <- buildSearchArgument(dataTypeMap, child, builder.startNot())
+          _ <- createBuilder(dataTypeMap, child, newBuilder, canPartialPushDownConjuncts = false)
+          negate <- createBuilder(dataTypeMap, child, builder.startNot(), false)
```
Comment (dbtsai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala]: Nit, can you explicitly call `canPartialPushDownConjuncts = false`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala [hunk 4]
```diff
@@ -90,32 +107,51 @@ private[orc] object OrcFilters extends Logging {
...
-        } yield rhs.end()
+        //
+        // Pushing one side of AND down is only safe to do at the top level or in the child
+        // AND before hitting NOT or OR conditions, and in this case, the unsupported predicate
+        // can be safely removed.
+        val leftBuilderOption = createBuilder(dataTypeMap, left,
+          newBuilder, canPartialPushDownConjuncts)
+        val rightBuilderOption =
+          createBuilder(dataTypeMap, right, newBuilder, canPartialPushDownConjuncts)
+        (leftBuilderOption, rightBuilderOption) match {
```
Comment (dbtsai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: ditto

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala [hunk 5]
```diff
@@ -90,32 +107,51 @@ private[orc] object OrcFilters extends Logging {
...
+            builder.startOr(), canPartialPushDownConjuncts = false)
+          rhs <- createBuilder(dataTypeMap, right, lhs, canPartialPushDownConjuncts = false)
         } yield rhs.end()
 
       case Not(child) =>
         for {
-          _ <- buildSearchArgument(dataTypeMap, child, newBuilder)
-          negate <- buildSearchArgument(dataTypeMap, child, builder.startNot())
+          _ <- createBuilder(dataTypeMap, child, newBuilder, canPartialPushDownConjuncts = false)
+          negate <- createBuilder(dataTypeMap, child, builder.startNot(), false)
```
Comment (dbtsai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala]: ditto

**Comment by dbtsai** [comment]
LGTM. Just some styling feedback. Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala]: Then it has to be two lines..I am OK with that too.

**Comment by dbtsai** [comment]
Merged into master. Thanks.

**Pull request closed without merging**